pkg/blob/controllerserver.go

```diff
@@ -786,11 +757,36 @@ func (d *Driver) DeleteBlobContainer(ctx context.Context, subsID, resourceGroupN
 }
 
 // copyBlobContainer copies source volume content into a destination volume
-func (d *Driver) copyBlobContainer(authAzcopyEnv []string, srcPath string, srcAccountSASToken string, dstPath string, dstAccountSASToken string, dstContainerName string) error {
+func (d *Driver) copyBlobContainer(ctx context.Context, req *csi.CreateVolumeRequest, dstAccountName string, dstAccountSasToken string, authAzcopyEnv []string, dstContainerName string, secretNamespace string, accountOptions *azure.AccountOptions, storageEndpointSuffix string) error {
+	var sourceVolumeID string
+	if req.GetVolumeContentSource() != nil && req.GetVolumeContentSource().GetVolume() != nil {
+		sourceVolumeID = req.GetVolumeContentSource().GetVolume().GetVolumeId()
 
-	if srcPath == "" || dstPath == "" || dstContainerName == "" {
-		return fmt.Errorf("srcPath(%s) or dstPath(%s) or dstContainerName(%s) is empty", srcPath, dstPath, dstContainerName)
 	}
+	resourceGroupName, srcAccountName, srcContainerName, _, _, err := GetContainerInfo(sourceVolumeID) //nolint:dogsled
+	if err != nil {
+		return status.Error(codes.NotFound, err.Error())
+	}
+	if dstAccountName == "" {
+		dstAccountName = srcAccountName
+	}
+	if srcAccountName == "" || srcContainerName == "" || dstContainerName == "" {
+		return fmt.Errorf("srcAccountName(%s) or srcContainerName(%s) or dstContainerName(%s) is empty", srcAccountName, srcContainerName, dstContainerName)
+	}
+	srcAccountSasToken := dstAccountSasToken
+	if srcAccountName != dstAccountName && dstAccountSasToken != "" {
+		srcAccountOptions := &azure.AccountOptions{
+			Name:                srcAccountName,
+			ResourceGroup:       accountOptions.ResourceGroup,
+			SubscriptionID:      accountOptions.SubscriptionID,
+			GetLatestAccountKey: accountOptions.GetLatestAccountKey,
+		}
+		if srcAccountSasToken, _, err = d.getAzcopyAuth(ctx, srcAccountName, "", storageEndpointSuffix, srcAccountOptions, nil, "", secretNamespace); err != nil {
```

Comment:
Does `getAzcopyAuth` work on a source volume for which the controller identity only has `Storage Blob Data Reader`.

Reply:
I think it may failed with `AuthorizationPermissionMismatch` error, as azcopy team mentioned, it needs `Storage Blob Data Contributor` role

Reply:
we don't know here, from intuition, `Storage Blob Data Reader` should work on source volume, need to verify, and if it failed, then fallback to use sas token

Reply:
I tested it to only give `Storage Blob Data Reader` in source blob, and it also return `AuthorizationPermissionMismatch` error from azcopy.
```
2024/08/01 08:46:05 INFO: [P#0-T#0] Starting transfer: Source "https://fuse256643cb981cf47f286.blob.core.windows.net/pvc-454d10fc-47c0-4d75-847c-48b4093c6261/outfile" Destination "https://fuse29375444a7b114aaf91.blob.core.windows.net/pvc-88f8b354-4d77-4097-a65b-0ecb93607e87/outfile". Specified chunk size 8388608
2024/08/01 08:46:05 ==> REQUEST/RESPONSE (Try=1/9.557829ms, OpTime=30.325026ms) -- RESPONSE STATUS CODE ERROR
   PUT https://fuse29375444a7b114aaf91.blob.core.windows.net/pvc-88f8b354-4d77-4097-a65b-0ecb93607e87/outfile
   Accept: application/xml
   Authorization: REDACTED
   Content-Length: 0
   User-Agent: AzCopy/10.25.1 azsdk-go-azblob/v1.3.1 (go1.22.4; linux)
   X-Ms-Client-Request-Id: 79584760-5d17-4747-4048-df4f78b38efe
   x-ms-blob-content-md5: tTzxIJei9Hch9GOVLgQ+WA==
   x-ms-blob-content-type: application/octet-stream
   x-ms-blob-type: BlockBlob
   x-ms-copy-source: https://fuse256643cb981cf47f286.blob.core.windows.net/pvc-454d10fc-47c0-4d75-847c-48b4093c6261/outfile
   x-ms-copy-source-authorization: REDACTED
   x-ms-version: 2023-08-03
   --------------------------------------------------------------------------------
   RESPONSE Status: 403 This request is not authorized to perform this operation using this permission.
   Content-Length: 279
   Content-Type: application/xml
   Date: Thu, 01 Aug 2024 08:46:05 GMT
   Server: Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0
   X-Ms-Client-Request-Id: 79584760-5d17-4747-4048-df4f78b38efe
   X-Ms-Error-Code: AuthorizationPermissionMismatch
   X-Ms-Request-Id: aa17dd77-001e-0022-5def-e3ef22000000
   X-Ms-Version: 2023-08-03
Response Details: <Code>AuthorizationPermissionMismatch</Code><Message>This request is not authorized to perform this operation using this permission. </Message>

2024/08/01 08:46:05 ERR: [P#0-T#0] COPYFAILED: https://fuse256643cb981cf47f286.blob.core.windows.net/pvc-454d10fc-47c0-4d75-847c-48b4093c6261/outfile : 403 : 403 This request is not authorized to perform this operation using this permission.. When Put Blob from URL. X-Ms-Request-Id: aa17dd77-001e-0022-5def-e3ef22000000
```

Reply:
It seems strange and doesn't match the public docs for [AzCopy](https://learn.microsoft.com/en-us/azure/storage/common/storage-use-azcopy-authorize-azure-active-directory#verify-role-assignments). I asked the team and they indicated that:

> This request appears to be a response from the destination account. It is not a source validation failure, which indicates that it's the actual request to the destination, not the source that's failing. Furthermore, it's AuthorizationPermissionMismatch, meaning that authorization has succeeded, but permissions are lacking.

In short-- sounds like perms are lacking on the destination.

Reply:
Here's a bicep POC that shows `azcopy` itself **does** work with only `Blob Data Reader` on the source and `Blob Data Contributor` on the destination.

The `blob-csi-driver` should support it as well for at least MSI scenarios.

```bicep
param stem string = uniqueString(resourceGroup().name)
param location string = resourceGroup().location
param azCliVersion string = '2.52.0'
param forceUpdateTag string = utcNow()

var blobOwnerRoleId = subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'b7e6dc6d-f1e8-4753-8033-0f276bb0955b')
var blobReaderRoleId = subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '2a2b9908-6ea1-4ae2-8e65-a410df84e7d1')
var blobContributorRoleId = subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'ba92f5b4-2d11-453d-a403-e96b0029c9fe')

resource srcAccount 'Microsoft.Storage/storageAccounts@2023-01-01' = {
  name: '${stem}src'
  location: location
  sku: {
    name: 'Standard_LRS'
  }
  kind: 'StorageV2'
  properties: {
    accessTier: 'Hot'
    allowSharedKeyAccess: false
    defaultToOAuthAuthentication: true
    isLocalUserEnabled: false
  }
}

resource dstAccount 'Microsoft.Storage/storageAccounts@2023-01-01' = {
  name: '${stem}dst'
  location: location
  sku: {
    name: 'Standard_LRS'
  }
  kind: 'StorageV2'
  properties: {
    accessTier: 'Hot'
    allowSharedKeyAccess: false
    defaultToOAuthAuthentication: true
    isLocalUserEnabled: false
  }
}

resource makeIdentity 'Microsoft.ManagedIdentity/userAssignedIdentities@2023-01-31' = {
  name: '${stem}-make-id'

  location: location
}

resource copyIdentity 'Microsoft.ManagedIdentity/userAssignedIdentities@2023-01-31' = {
  name: '${stem}-copy-id'
  location: location
}

resource makeSourceBlobOwner 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
  name: guid(makeIdentity.id, blobOwnerRoleId, srcAccount.id)
  properties: {
    roleDefinitionId: blobOwnerRoleId
    principalId: makeIdentity.properties.principalId
  }
  scope: srcAccount
}

resource copySourceBlobReader 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
  name: guid(copyIdentity.id, blobReaderRoleId, srcAccount.id)
  properties: {
    roleDefinitionId: blobReaderRoleId
    principalId: copyIdentity.properties.principalId
  }
  scope: srcAccount
}


resource copyDestinationBlobContributor 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
  name: guid(copyIdentity.id, blobContributorRoleId, dstAccount.id)
  properties: {
    roleDefinitionId: blobContributorRoleId
    principalId: copyIdentity.properties.principalId
  }
  scope: dstAccount
}

resource makeScript 'Microsoft.Resources/deploymentScripts@2023-08-01' = {
  name: '${stem}-make-script'
  location: location
  kind: 'AzureCLI'
  identity: {
    type: 'UserAssigned'
    userAssignedIdentities: {
      '${makeIdentity.id}': {}
    }
  }
  properties: {
    azCliVersion: azCliVersion
    forceUpdateTag: forceUpdateTag
    environmentVariables: [
      {
        name: 'AZCOPY_AUTO_LOGIN_TYPE'
        value: 'MSI'
      }
      {
        name: 'SRCACCOUNT'
        value: srcAccount.properties.primaryEndpoints.blob
      }
      {
        name: 'SRCCONTAINER'
        value: 'data'
      }
    ]
    scriptContent: '${installScriptContent}\n${makeScriptContent}'
    retentionInterval: 'PT1H'
  }
  dependsOn: [
    makeSourceBlobOwner
  ]
}

resource copyScript 'Microsoft.Resources/deploymentScripts@2023-08-01' = {
  name: '${stem}-copy-script'
  location: location
  kind: 'AzureCLI'
  identity: {
    type: 'UserAssigned'
    userAssignedIdentities: {
      '${copyIdentity.id}': {}
    }
  }
  properties: {
    azCliVersion: azCliVersion
    forceUpdateTag: forceUpdateTag
    environmentVariables: [
      {
        name: 'AZCOPY_AUTO_LOGIN_TYPE'
        value: 'MSI'
      }
      {
        name: 'SRCACCOUNT'
        value: srcAccount.properties.primaryEndpoints.blob
      }
      {
        name: 'SRCCONTAINER'
        value: 'data'
      }
      {
        name: 'DSTACCOUNT'
        value: dstAccount.properties.primaryEndpoints.blob
      }
      {
        name: 'DSTCONTAINER'
        value: 'data'
      }
    ]
    scriptContent: '${installScriptContent}\n${copyScriptContent}'
    retentionInterval: 'PT1H'
  }
  dependsOn: [
    copySourceBlobReader
    copyDestinationBlobContributor
    makeScript
  ]
}

var installScriptContent = '''
cd /tmp
mkdir bin
apk add -q --no-interactive gcompat
wget -q https://aka.ms/downloadazcopy-v10-linux
tar -xf downloadazcopy-v10-linux
cp -vf ./azcopy_linux_amd64_*/azcopy bin/
rm -rf ./azcopy_linux_amd64_* downloadazcopy-v10-linux
'''

var makeScriptContent = '''
mkdir data
date > data/0.txt
date > data/1.txt
date > data/3.txt
date > data/4.txt
./bin/azcopy jobs list
./bin/azcopy list ${SRCACCOUNT}
./bin/azcopy make ${SRCACCOUNT}${SRCCONTAINER}
./bin/azcopy copy data ${SRCACCOUNT}${SRCCONTAINER} --recursive
 '''

var copyScriptContent = '''
./bin/azcopy jobs list
./bin/azcopy list ${SRCACCOUNT}
./bin/azcopy list ${DSTACCOUNT}
./bin/azcopy copy ${SRCACCOUNT}${SRCCONTAINER} ${DSTACCOUNT}${DSTCONTAINER} --recursive --check-length=false --s2s-preserve-access-tier=false
'''
```

Reply:
thanks for the verification, we could address this in the doc or other PR